### ISSUE TITLE
ISR: Invalidate cache issue

### ISSUE THREAD

**Issue by artemserdyuk** [rx-angular/rx-angular#1748]
**When we make an API request, we noticed that the library does not invalidate the cache.**

[(https://www.rx-angular.io/docs/isr/on-demand-revalidation)]
```
server.post(
    '/api/invalidate',
    async (req, res) => await isr.invalidate(req, res)
  `);`
```
When we compare the original files from [(https://github.com/rx-angular/rx-angular/blob/main/libs/isr/server)] with the files installed in **node_modules**, we noticed that there is a difference in the **renderUrl()** method in the installed files in node_modules (**rx-angular-isr-server.mjs**), compared to the file from [(https://github.com/rx-angular/rx-angular/blob/main/libs/isr/server/src/utils/render-url.ts)].

![1](https://github.com/user-attachments/assets/f60aab8d-b19e-42f3-9e48-4bf354339cdf)
![2](https://github.com/user-attachments/assets/651083a7-64e2-4567-9618-cb0513334af9)

We suspect that there is an error in the **renderUrl()** method in the installed node_modules file **rx-angular-isr-server.mjs**. See the screenshots.

**Comment by maxisam** [comment]
I don't see this with my build files. I check both fesm2022 and esm2022. The code order is the same with the source code.

**Comment by eneajaho** [comment]
@artemserdyuk 
Can you delete node_modules and install again, and also tell the ng version? 
Thanks

**Comment by arturober** [comment]
I can confirm this bug only when I try to invalidate the '/' route. I get a confirmation that it's been invalidated in the response and the docker logs (_Urls: / were regenerated!_), however I always get the old content. Other urls work as expected, or so it seems to me.

**Comment by arturober** [comment]
This might be useful. Even if I **remove the revalidate** attribute for the '/' route, it stills stores it in cache. This is the last part of the HTML I get from the server:
```html
<script id="isr-state" type="application/json">{"revalidate":null}</script></body></html><!-- 
🚀 ISR: Served from cache! 
⌛ Last updated: 2024-09-30 09:57:56.  
-->
```
Like before, it stills never invalidates this route. I have to restart the service to delete the cache...

**Comment by maxisam** [comment]
That message is misleading actually. I just noticed it last weekend and change it locally in our system as well 

It runs every time when the page is generated from SSR, so I changed it to "🚀 ISR: Served from SSR!"

You should check the last updated time to see when it was generated so you can tell if it was from cache or not

**Comment by arturober** [comment]
You're right about the time. It shows always the current time so it shouldn't come from cache. But, the content (HTML) is not updated, like it's been stored in cache the first time the route '/' is requested and gets it from cache every time... It's the only route where it happens. Other routes are invalidated correctly and return the updated content, as they should.

These are the route files in case it helps:

**app.routes.ts**
```typescript
import { Routes } from '@angular/router';

export const appRoutes: Routes = [
  {
    path: '',
    title: 'Cursos de Programación | FullStackPro.es',
    loadComponent: () =>
      import('./home/home.component').then((m) => m.HomeComponent),
    pathMatch: 'full',
    // data: {
    //   revalidate: 0, // 👈 Add the revalidate key
    // },
  },
  {
    path: 'courses',
    loadChildren: () =>
      import('./course/course.routes').then((m) => m.courseRoutes),
  },
  {
    path: 'users',
    loadChildren: () =>
      import('./users/users.routes').then((m) => m.usersRoutes),
  },
  { path: '**', redirectTo: '', pathMatch: 'full' },
];
```
**course/course.routes.ts**
```typescript
export const courseRoutes: Routes = [
  { path: '', redirectTo: '/', pathMatch: 'full' },
  {
    path: ':course/:section',
    loadComponent: () => import('./course.component').then(m => m.CourseComponent),
    data: {
      revalidate: 0, // 👈 Add the revalidate key
    },
  },
  {
    path: ':course',
    loadComponent: () => import('./course.component').then(m => m.CourseComponent),
    data: {
      revalidate: 0, // 👈 Add the revalidate key
    },
  },
];
```

/users routes are private and therefore not cached

**Comment by maxisam** [comment]
if you don't have revalidate, it won't cache and SSR every time. So if you have different content from backend, it will change everytime. I think you can create a test app to test this simply. Just fork this repo and add some code here

https://github.com/rx-angular/rx-angular/tree/main/apps/ssr-isr

like get current date time and show it on the page, so everytime it should render differently.

**Comment by maxisam** [comment]
nay, you don't need to record video, I think I know what you meant. I just want you to create a small test repo if you have time. 

and not using web api but just something like current datetime for updating the page. 

My suspect is somehow SSR rendering cache your http request. So everytime it renders, it got the same data (http web api cache). That is why I want you to test it with sometime not from the web api.

If you can check your web api server and disable cache, maybe it will work.

**Comment by arturober** [comment]
I've reproduced the bug as you said. The date is not updated in the HTML I get from the server. I have not set any route (no revalidate anywhere). This is the repository for testing this:

[https://github.com/arturober/isr-test](https://github.com/arturober/isr-test)

Test with **ng build && node dist/isr-test/server/server.mjs**

Try just http://localhost:4000 and the bug is there, but add any path to the route (ex: http://localhost:4000/hello) and it works like it should.

Maybe I'm doing domething wrong, but it's strange everything works ok for routes other than empty route.

**Comment by maxisam** [comment]
ah... It does update for me. 

first load
![image](https://github.com/user-attachments/assets/007d732f-14e8-46fe-a418-7804f93d86d4)

ctrl+f5
![image](https://github.com/user-attachments/assets/47f6f842-f36b-43d4-8574-4810b3577079)

Do I misunderstand?

**Comment by arturober** [comment]
Look at the date inside the <body> element ;)

**Comment by maxisam** [comment]
you set prerender: true in your `angular.json`

it should be 

```json
            "prerender": {
              "discoverRoutes": false
            },
```

**Comment by arturober** [comment]
My bad, that was the problem. I didn't expect that SGR was enabled by default (for static routes) in Angular 18 when enabling SSR. I don't think that's a good idea for a default configuration... Thank you!

I would add this step in the documentation, because I fear I won't be the only one with this problem as this is the default configuration Angular set when enabling SSR/SGR (in newest versions at least, SGR is relatively new in Angular).

Keep up the good work!, this library is a must have. I don't understand why Angular doesn't support ISR natively yet. Thank you again.

**Comment by maxisam** [comment]
Well, I think they just can't wait to show SSG finally landed in Angular lol Welcome to update the doc 😄
